File: kotlin/kotlin-1.6.10/src/jmh/kotlin/dev/nulli/jvm_benchmarks/kotlin_1_6_10/execution/AbstractLoopExecution.kt
package dev.nulli.jvm_benchmarks.kotlin_1_6_10.execution

import org.openjdk.jmh.annotations.*
import kotlin.random.Random

@State(Scope.Benchmark)
abstract class AbstractLoopExecution<E, L>(private val valuesFunc: (Random, Int) -> L, private val modifierFunc: (Random) -> E) {

    @Param(
        "100",
        "1000", "5000",
        "1000000", "2000000", "3000000", "5000000", "7000000",
        "10000000", "20000000", "30000000", "50000000", "70000000",
        "100000000"
    )
    protected var iterations: Int = 0

    var values: L? = null
    var modifier: E? = null

    @Setup(Level.Iteration)
    fun setup() {
        this.values = null
        this.values = this.valuesFunc.invoke(Random, this.iterations)
        this.modifier = this.modifierFunc.invoke(Random)
    }
}

File: scala/scala-2.13.8/src/jmh/scala/dev/nulli/jvm_benchmarks/scala_2_13_8/execution/AbstractLoopExecution.scala
package dev.nulli.jvm_benchmarks.scala_2_13_8.execution

import org.openjdk.jmh.annotations.{Level, Param, Scope, Setup, State}

import scala.util.Random

@State(Scope.Benchmark)
abstract class AbstractLoopExecution[E, L](private val valuesFunc: (Random, Int) => L, private val modifierFunc: Random => E) {

    @Param(Array(
        "100",
        "1000", "5000",
        "1000000", "2000000", "3000000", "5000000", "7000000",
        "10000000", "20000000", "30000000", "50000000", "70000000",
        "100000000"
    ))
    protected var iterations: Int = 0

    var values: Option[L] = None
    var modifier: Option[E] = None

    @Setup(Level.Iteration)
    def setup(): Unit = {
        this.values = None
        this.values = Some(this.valuesFunc.apply(Random, this.iterations))
        this.modifier = Some(this.modifierFunc.apply(Random))
    }
}
